File: MTWDM iOS Xamarin/AppSQLite/AppSQLite/ViewController.cs
using System;
using UIKit;
using System.Collections.Generic;
using Entidades;
using Foundation;
using System.Linq;

namespace AppSQLite
{
    public partial class ViewController : UIViewController
    {
        Utilidades objUtilidades;

        List<Empresa> lista;

        Conexion objConexion;

        NSObject observer;

        UISearchController searchController;

        protected ViewController(IntPtr handle) : base(handle)
        {
            // Note: this .ctor should not contain any initialization logic.
        }

        public override void ViewDidLoad()
        {
            base.ViewDidLoad();
            // Perform any additional setup after loading the view, typically from a nib.
            objUtilidades = new Utilidades(this);

            objConexion = new Conexion();

            lista = new List<Empresa>();

            //objUtilidades.MessageBox("Ejemplo","Mensaje","Alert");
            //objUtilidades.MessageBox("Ejemplo", "Mensaje", "Sheet");

            //Lenar la UITableView
            LlenarTabla();

            observer =
                NSNotificationCenter.DefaultCenter.AddObserver((NSString)"NSUserDefaultsDidChangeNotification",
                                                               UpdateSettings);
                                          
            btnEditar.Clicked += delegate {

                if (!Tabla.Editing) {

                    btnEditar.Title = "Listo";
                    Tabla.SetEditing(true, true);
                } else {

                    btnEditar.Title = "Editar";
                    Tabla.SetEditing(false, true);
                }

            };

        }

        public override void ViewDidUnload()
        {
            base.ViewDidUnload();

            if (observer != null)
            {
                NSNotificationCenter.DefaultCenter.RemoveObserver(observer);
                observer = null;
            }
        }

        private void UpdateSettings(NSNotification obj)
        {
            habilitarSearchBarController();
        }

        void habilitarSearchBarController(){

            this.Title = Settings.tituloTabla; //"Lista";

            if (UIDevice.CurrentDevice.CheckSystemVersion(11, 0))
            {

                searchController = new UISearchController(searchResultsController: null)
                {
                    HidesNavigationBarDuringPresentation = true,
                    DimsBackgroundDuringPresentation = false
                };

                searchController.SearchBar.SearchBarStyle = UISearchBarStyle.Minimal;

                searchController.SearchBar.Placeholder = "Buscar";

                searchController.SearchBar.TextChanged += SearchBar_TextChanged;

                searchController.SearchBar.CancelButtonClicked += (sender, e) => {

                    LlenarTabla();
                };

                if (Settings.habilitarbusqueda)
                {
                    searchController.SearchBar.Hidden = false;
                    NavigationItem.SearchController = searchController;
                }
                else
                {
                    searchController.SearchBar.Hidden = true;
                    NavigationItem.SearchController = null;
                }

                NavigationItem.HidesSearchBarWhenScrolling = false;

                NavigationController.NavigationBar.PrefersLargeTitles = true;


            }
        }

        void SearchBar_TextChanged(object sender, UISearchBarTextChangedEventArgs e)
        {
            string capturaDato = e.SearchText;

            LlenarTabla(capturaDato);
        }

        public override void ViewWillAppear(bool animated)
        {
            base.ViewWillAppear(animated);

            UpdateSettings(null);

            var Empresas = objConexion.conexion.Table<Empresa>();

            lista.Clear();
            foreach (var fila in Empresas)
            {
                lista.Add(new Empresa
                {
                    Domicilio = fila.Domicilio,
                    Nombre = fila.Nombre,
                    RFC = fila.RFC,
                    RepresentanteLegal = fila.RepresentanteLegal
                });
            }

            Tabla.Source = null;
            Tabla.Source = new DatosTableSource(lista,this);
            Tabla.ReloadData();
        }

        public override void DidReceiveMemoryWarning()
        {
            base.DidReceiveMemoryWarning();
        }

        public void LlenarTabla(string busqueda = null)
        {

            lista.Clear();

            var Empresas = objConexion.conexion.Table<Empresa>();

            foreach (var fila in Empresas)
            {
                lista.Add(new Empresa
                {
                    Domicilio = fila.Domicilio,
                    Nombre = fila.Nombre,
                    RFC = fila.RFC,
                    RepresentanteLegal = fila.RepresentanteLegal
                });
            }

      

            if (string.IsNullOrEmpty(busqueda))
            {

                Tabla.Source = null;
                Tabla.Source = new DatosTableSource(lista, this);
                Tabla.ReloadData();

            }
            else
            {
                var l = lista.Where(r => r.Nombre.Contains(busqueda)).ToList();
                Tabla.Source = null;
                Tabla.Source = new DatosTableSource(l, this);
                Tabla.ReloadData();

            }


        }

    }


    public class DatosTableSource : UITableViewSource
    {
        List<Empresa> _datos;

        Conexion objConexion;

        Utilidades objUtilidades;

        UIViewController viewController;

        public DatosTableSource(List<Empresa> datos, UIViewController controller)
        {
            _datos = datos;

            objConexion = new Conexion();

            viewController = controller;

            objUtilidades = new Utilidades(viewController);
        }

        public override bool CanEditRow(UITableView tableView, NSIndexPath indexPath)
        {
            return true; // return false if you wish to disable editing for a specific indexPath or for all rows
        }
      
        public override UITableViewCellEditingStyle EditingStyleForRow(UITableView tableView, NSIndexPath indexPath)
        {
            return UITableViewCellEditingStyle.Delete; // this example doesn't support Insert
        }

        public override string TitleForDeleteConfirmation(UITableView tableView, NSIndexPath indexPath)
        {
            return "Eliminar (" + _datos[indexPath.Row].Nombre + ")";
        }

        public override void CommitEditingStyle(UITableView tableView, 
                                                UITableViewCellEditingStyle editingStyle,
                                                NSIndexPath indexPath)
        {
            switch (editingStyle)
            {
                case UITableViewCellEditingStyle.Delete:

                    objConexion.conexion.Delete(_datos[indexPath.Row]);


                    //SE ELIMINA LA FILA DEL DATASOURCE
                    _datos.RemoveAt(indexPath.Row);

                    //SE ELIMINA  LA FILA DE LA TABLA
                    tableView.DeleteRows(new NSIndexPath[] { indexPath },
                                         UITableViewRowAnimation.Fade);
                    break;
                case UITableViewCellEditingStyle.None:
                    Console.WriteLine("CommitEditingStyle:none called");
                    break;
                default: break;
            }
        }

        public override nint RowsInSection(UITableView tableview, nint section)
        {
            return _datos.Count;
        }

        public override UITableViewCell GetCell(UITableView tableView,
                                                NSIndexPath indexPath)
        {
            UITableViewCell cell = tableView.DequeueReusableCell("TableCell");
            if (cell == null)
            {
                cell = new UITableViewCell(UITableViewCellStyle.Subtitle,
                                           "TableCell");
            }
            //show title
            cell.TextLabel.Text = _datos[indexPath.Row].Nombre;

            cell.DetailTextLabel.Text = _datos[indexPath.Row].RFC;

            //add accessory
            cell.Accessory = UITableViewCellAccessory.DetailButton;

            return cell;
        }


        public override void AccessoryButtonTapped(UITableView tableView, NSIndexPath indexPath)
        {

            var controller = viewController.Storyboard.InstantiateViewController("EditarViewController") as EditarViewController;

            controller.objEmpresa = _datos[indexPath.Row];

            viewController.NavigationController.ModalPresentationStyle = UIModalPresentationStyle.Custom;

            viewController.NavigationController.PushViewController(controller,true);
        }

        public override void RowSelected(UITableView tableView, NSIndexPath indexPath)
        {
            
            tableView.DeselectRow(indexPath, false);


            string msg = string.Format("Empresa:{0}\nDomicilio:{1}\nRFC:{2}\nRepresentante Legal:{3}",
                                                _datos[indexPath.Row].Nombre,
                                                _datos[indexPath.Row].Domicilio,
                                                _datos[indexPath.Row].RFC,
                                                _datos[indexPath.Row].RepresentanteLegal);


            objUtilidades.MessageBox("Datos", msg, "Sheet");
        }

    }
}

File: MTWDM iOS Swift /appSQLITE/appSQLITE/ViewController.swift
//
//  ViewController.swift
//  appSQLITE
//
//  Created by Felipe Hernandez on 12/7/18.
//  Copyright © 2018 Felipe Hernandez. All rights reserved.
//

import UIKit
import CoreData

class ViewController: UIViewController {

    
    @IBOutlet weak var btnEditar: UIBarButtonItem!
    
    @IBOutlet weak var Tabla: UITableView!
    
    var objUtilidades: Utilidades?
    
    var searchController : UISearchController?
    
    
    var objPersona: Persona?
    
    var listPersonas = [NSManagedObject]()
    
    // UIViewController viewController;
    var viewController : UIViewController?
    
    @IBAction func onEditar(_ sender: Any) {
        
        if Tabla.isEditing {
            btnEditar.title = "Editar"
            Tabla.setEditing(false, animated: true)
        }else {
            btnEditar.title = "Listo"
            Tabla.setEditing(true, animated: true)
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        objUtilidades =  Utilidades(_viewController: self)
        
        //objUtilidades?.MessageBox(Titulo: "Prueba", Mensaje: "Prueba", tipo: tipoMsg.Alert)
        //objUtilidades?.MessageBox(Titulo: "Prueba", Mensaje: "Prueba", tipo: tipoMsg.Sheet)
        
        viewController =  self
        
        habilitarSearchBarController()
        
    }
    
    
    func LlenarTabla(dato:String?) {
        
        let appDelegate = UIApplication.shared.delegate  as! AppDelegate
        let managedContext = appDelegate.persistentContainer.viewContext
        
        let fecthRequest :  NSFetchRequest<Persona> = Persona.fetchRequest()
        
        if dato != "" {
    
            let resultPredicate = NSPredicate(format: "nombre contains[c] %@", dato!)
            fecthRequest.predicate = resultPredicate
            
        }
        

        do {
            let resultados  =  try managedContext.fetch(fecthRequest)
            
            listPersonas = resultados as [NSManagedObject]
            
        } catch let error as NSError {
            print("Error \(error) \(error.userInfo)")
        }
        
        Tabla.reloadData()
    }
    
    func habilitarSearchBarController()  {
        
        if #available(iOS 11.0, *) {
            
            searchController = UISearchController(searchResultsController: nil)
            
            searchController?.hidesNavigationBarDuringPresentation = true
            searchController?.dimsBackgroundDuringPresentation =  false
            
            searchController?.searchBar.searchBarStyle = .minimal
            
            searchController?.searchBar.placeholder = "Buscar"
            
            //Habilitar la busqueda
            searchController?.searchBar.isHidden =  false
            navigationItem.searchController  =  searchController
            
            navigationItem.hidesSearchBarWhenScrolling =  false
            navigationController?.navigationBar.prefersLargeTitles =  true
            
            searchController?.searchBar.delegate =  self
            
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        LlenarTabla(dato: "")
    }

}



extension ViewController : UISearchBarDelegate {
    
    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
        
        let capturaDato = searchText;
        
        print("dato escrito \(capturaDato)")
        
        LlenarTabla(dato: capturaDato);
    }
    
    func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
        
        LlenarTabla(dato: "")
    }
    
}


extension ViewController: UITableViewDataSource, UITableViewDelegate {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return listPersonas.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let p: Persona? = listPersonas[indexPath.row] as? Persona
        
        
        var cell =  tableView.dequeueReusableCell(withIdentifier: "cellPersona")
        
        if  cell != nil {
            
            cell =  UITableViewCell(style: .subtitle, reuseIdentifier: "cellPersona")
            
            cell!.textLabel?.text  = p?.nombre
            
            cell!.detailTextLabel?.text = p?.rfc
            
            cell!.accessoryType = .detailButton
            
        }
        
        return cell!
        
        
    }
    
    func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return true
    }
    
    func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCell.EditingStyle {
        return UITableViewCell.EditingStyle.delete
    }
    
    func tableView(_ tableView: UITableView, titleForDeleteConfirmationButtonForRowAt indexPath: IndexPath) -> String? {
        return "Eliminar"
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
       
        let p: Persona? = listPersonas[indexPath.row] as? Persona
        
        let msg  = "Empresa:\((p!.nombre)!) \nDomicilio:\((p!.domicilio)!)\nRFC:\((p!.rfc)!)\nRepresentante Legal:\((p!.representanteLegal)!)"
        
        objUtilidades?.MessageBox(Titulo: "Datos", Mensaje: msg, tipo: tipoMsg.Sheet)
        
        tableView.deselectRow(at: indexPath, animated: true)
    }
    
    func tableView(_ tableView: UITableView, accessoryButtonTappedForRowWith indexPath: IndexPath) {
      
        let p: Persona? = listPersonas[indexPath.row] as? Persona
        
        let controller = viewController!.storyboard?.instantiateViewController(withIdentifier: "EditarViewController") as! EditarViewController
        
        controller.p = p
        
        viewController!.navigationController?.modalPresentationStyle = .custom
        
        viewController!.navigationController?.pushViewController(controller, animated: true)
    }
    
    func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        
        switch editingStyle {
        case .delete:
            
            let appDelegate = UIApplication.shared.delegate  as! AppDelegate
            let managedContext = appDelegate.persistentContainer.viewContext
            
            let p: Persona? = listPersonas[indexPath.row] as? Persona
        
            managedContext.delete(p!)
            
            listPersonas.remove(at: indexPath.row)

            tableView.deleteRows(at: [indexPath], with: .fade)
            
            do {
                try managedContext.save()
                LlenarTabla(dato: "")
            }catch let error as NSError {
                 print("Error \(error) \(error.userInfo)")
            }
            
        case .none:
            print("No aplica")
        default:
            break
        }
    }
    
}

